In [1]:
# Initial setup by importing needed modules
import pandas as pd
import numpy as np
import datetime as dt
pd.options.display.max_columns = None
pd.options.display.max_rows=None

In [2]:
# Read off data and put that into a panda dataframe
lend = pd.read_csv('data/LoanStats3a.csv', header=1, low_memory=False)

Understanding data structure, checking shape and data type

In [3]:
print(lend.shape)
lend.info()

(42538, 144)
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 42538 entries, 0 to 42537
Columns: 144 entries, id to settlement_term
dtypes: float64(115), object(29)
memory usage: 46.7+ MB


There are 144 columns, 115 of those are floats data type. We will look at float data type columns first. 

In [4]:
lend.select_dtypes(include='float').head()

member_id  loan_amnt  funded_amnt  funded_amnt_inv  installment  \
0        NaN     5000.0       5000.0           4975.0       162.87   
1        NaN     2500.0       2500.0           2500.0        59.83   
2        NaN     2400.0       2400.0           2400.0        84.33   
3        NaN    10000.0      10000.0          10000.0       339.31   
4        NaN     3000.0       3000.0           3000.0        67.79   

   annual_inc  url    dti  delinq_2yrs  inq_last_6mths  \
0     24000.0  NaN  27.65          0.0             1.0   
1     30000.0  NaN   1.00          0.0             5.0   
2     12252.0  NaN   8.72          0.0             2.0   
3     49200.0  NaN  20.00          0.0             1.0   
4     80000.0  NaN  17.94          0.0             0.0   

   mths_since_last_delinq  mths_since_last_record  open_acc  pub_rec  \
0                     NaN                     NaN       3.0      0.0   
1                     NaN                     NaN       3.0      0.0   
2                     NaN                     NaN       2.0      0.0   
3                    35.0                     NaN      10.0      0.0   
4                    38.0                     NaN      15.0      0.0   

   revol_bal  total_acc  out_prncp  out_prncp_inv   total_pymnt  \
0    13648.0        9.0        0.0            0.0   5863.155187   
1     1687.0        4.0        0.0            0.0   1014.530000   
2     2956.0       10.0        0.0            0.0   3005.666844   
3     5598.0       37.0        0.0            0.0  12231.890000   
4    27783.0       38.0        0.0            0.0   4066.908161   

   total_pymnt_inv  total_rec_prncp  total_rec_int  total_rec_late_fee  \
0          5833.84          5000.00         863.16                0.00   
1          1014.53           456.46         435.17                0.00   
2          3005.67          2400.00         605.67                0.00   
3         12231.89         10000.00        2214.92               16.97   
4          4066.91          3000.00        1066.91                0.00   

   recoveries  collection_recovery_fee  last_pymnt_amnt  \
0         0.0                     0.00           171.62   
1       122.9                     1.11           119.66   
2         0.0                     0.00           649.91   
3         0.0                     0.00           357.48   
4         0.0                     0.00            67.30   

   collections_12_mths_ex_med  mths_since_last_major_derog  policy_code  \
0                         0.0                          NaN          1.0   
1                         0.0                          NaN          1.0   
2                         0.0                          NaN          1.0   
3                         0.0                          NaN          1.0   
4                         0.0                          NaN          1.0   

   annual_inc_joint  dti_joint  verification_status_joint  acc_now_delinq  \
0               NaN        NaN                        NaN             0.0   
1               NaN        NaN                        NaN             0.0   
2               NaN        NaN                        NaN             0.0   
3               NaN        NaN                        NaN             0.0   
4               NaN        NaN                        NaN             0.0   

   tot_coll_amt  tot_cur_bal  open_acc_6m  open_act_il  open_il_12m  \
0           NaN          NaN          NaN          NaN          NaN   
1           NaN          NaN          NaN          NaN          NaN   
2           NaN          NaN          NaN          NaN          NaN   
3           NaN          NaN          NaN          NaN          NaN   
4           NaN          NaN          NaN          NaN          NaN   

   open_il_24m  mths_since_rcnt_il  total_bal_il  il_util  open_rv_12m  \
0          NaN                 NaN           NaN      NaN          NaN   
1          NaN                 NaN           NaN      NaN          NaN   
2          NaN                 NaN

It looks like they are information on describing the loans itself. Given the problem is to predict the loan default/paid off chances at the point of applications. Details about the loans (such as terms, interest rates, hardship_length, etc.) are unknown upfront. Those columns will not be used in the machine learning analysis.

Reviewing the 29 object columns. Some of the columns actually have date time or percentage values, but panda cannot recognize them in the initial load. Those columns data will need further cleansing

In [5]:
lend.select_dtypes(exclude='float').head()

id        term int_rate grade sub_grade                 emp_title  \
0  NaN   36 months   10.65%     B        B2                       NaN   
1  NaN   60 months   15.27%     C        C4                     Ryder   
2  NaN   36 months   15.96%     C        C5                       NaN   
3  NaN   36 months   13.49%     C        C1       AIR RESOURCES BOARD   
4  NaN   60 months   12.69%     B        B5  University Medical Group   

  emp_length home_ownership verification_status   issue_d  loan_status  \
0  10+ years           RENT            Verified  Dec-2011   Fully Paid   
1   < 1 year           RENT     Source Verified  Dec-2011  Charged Off   
2  10+ years           RENT        Not Verified  Dec-2011   Fully Paid   
3  10+ years           RENT     Source Verified  Dec-2011   Fully Paid   
4     1 year           RENT     Source Verified  Dec-2011   Fully Paid   

  pymnt_plan                                               desc  \
0          n    Borrower added on 12/22/11 > I need to upgra...   
1          n    Borrower added on 12/22/11 > I plan to use t...   
2          n                                                NaN   
3          n    Borrower added on 12/21/11 > to pay for prop...   
4          n    Borrower added on 12/21/11 > I plan on combi...   

          purpose                 title zip_code addr_state earliest_cr_line  \
0     credit_card              Computer    860xx         AZ         Jan-1985   
1             car                  bike    309xx         GA         Apr-1999   
2  small_business  real estate business    606xx         IL         Nov-2001   
3           other              personel    917xx         CA         Feb-1996   
4           other              Personal    972xx         OR         Jan-1996   

  revol_util initial_list_status last_pymnt_d next_pymnt_d last_credit_pull_d  \
0      83.7%                   f     Jan-2015          NaN           Dec-2018   
1       9.4%                   f     Apr-2013          NaN           Oct-2016   
2      98.5%                   f     Jun-2014          NaN           Jun-2017   
3        21%                   f     Jan-2015          NaN           Apr-2016   
4      53.9%                   f     Jan-2017          NaN           Apr-2018   

  application_type hardship_flag debt_settlement_flag  \
0       Individual             N                    N   
1       Individual             N                    N   
2       Individual             N                    N   
3       Individual             N                    N   
4       Individual             N                    N   

  debt_settlement_flag_date settlement_status settlement_date  
0                       NaN               NaN             NaN  
1                       NaN               NaN             NaN  
2                       NaN               NaN             NaN  
3                       NaN               NaN             NaN  
4                       NaN               NaN             NaN

Drop data will large amount of missing data. We will use isnull and sum function to check missing for each columns. 

In [6]:
lend.isnull().sum()

id                                            42535
member_id                                     42538
loan_amnt                                         3
funded_amnt                                       3
funded_amnt_inv                                   3
term                                              3
int_rate                                          3
installment                                       3
grade                                             3
sub_grade                                         3
emp_title                                      2629
emp_length                                     1115
home_ownership                                    3
annual_inc                                        7
verification_status                               3
issue_d                                           3
loan_status                                       3
pymnt_plan                                        3
url                                           42538
desc        

The dataset has a total of 42K row of data. Some columns almost have no data at all, and wouldn't be useful. By scanning the missing records range, I will make the decision to drop columns that have >40K missing records. Loan_amnt has three missing records (which are blank rows), we will drop those as well.

In [7]:
lend_drop = lend.dropna(axis=1, how='all', thresh=40000).dropna(subset=['loan_amnt'])
lend_drop.shape

(42535, 51)

For the float type columns, we will replace missing values are equivalent to 0

In [8]:
# Suppress output since it will take forever for the display
lend_drop.select_dtypes(include='float').fillna(0);

For the object columns, further data type conversion and filling missing values will be need. We will focus on issue_d, emp_length, earliest_cr_line, revol_util and title. Other columns are either unknown variables for machine learning or no data type conversions are needed.

In [9]:
lend_drop.select_dtypes(exclude='float').isnull().sum()

term                       0
int_rate                   0
grade                      0
sub_grade                  0
emp_length              1112
home_ownership             0
verification_status        0
issue_d                    0
loan_status                0
pymnt_plan                 0
purpose                    0
title                     13
zip_code                   0
addr_state                 0
earliest_cr_line          29
revol_util                90
initial_list_status        0
last_pymnt_d              83
last_credit_pull_d         4
application_type           0
hardship_flag              0
debt_settlement_flag       0
dtype: int64

Loan issues date - Convert into datetime value

In [10]:
#use assign function to avoid security type warining
lend_drop = lend_drop.assign(loan_start_d=pd.to_datetime(lend_drop.issue_d, format='%b-%Y'))

Employment duration
- convert from text to numeric value. 
- clean up na value, assume <1 year if info is not provided (currently unemployment)

In [11]:
lend_drop['yr_emp'] = lend_drop['emp_length'].str[0:2].replace('< ',0).astype('float')

Credit History
- Convert earliest_cr_line into from text to datetime values
- Replace missing values with zero
- Extract year and convert to categorical values to have higher level grouping for this features

In [12]:
lend_drop.earliest_cr_line = pd.to_datetime(lend_drop.earliest_cr_line, format='%b-%Y')
lend_drop['yr_credit']= (2011-lend_drop.earliest_cr_line.dt.year).fillna(0)

Credit line % usage
- Convert from string to decimal place

In [13]:
lend_drop['revol_util_dec'] = lend_drop['revol_util'].str.replace(r'%',r'0').astype('float')/100

loan_status will serve as target for this analysis

In [14]:
lend_drop.loan_status.value_counts()

Fully Paid                                             34116
Charged Off                                             5670
Does not meet the credit policy. Status:Fully Paid      1988
Does not meet the credit policy. Status:Charged Off      761
Name: loan_status, dtype: int64

Convert text value feature into numerics for model training purposes. Create a dictionary to convert loan status into 0 (default) or 1 (Paid)

Per research through discussion forum, "Dose not meet the credit policy' are loans that had issued before Lending Club update credit policy. For the purpose of this study, we will treat them as "Fully Paid" or "Charged Off', regardless of credit policy compliance.

In [15]:
loan_dict = {'Fully Paid': 1, 'Charged Off': 0, 'Does not meet the credit policy. Status:Fully Paid': 1, 'Does not meet the credit policy. Status:Charged Off': 0 }

Create a target column by mapping loan status to dictionary,

In [16]:
#use assign function to avoid security type warining
lend_drop = lend_drop.assign(target=lend_drop.loan_status.map(loan_dict))

Save cleaned data set into a csv file for quicker access.

In [17]:
lend_drop.to_csv('data/lending_clean.csv', index=False)